systemd/coreos-metadata-sshkeys@.service

```diff
@@ -0,0 +1,11 @@
+[Unit]
+Description=CoreOS Metadata Agent (SSH Keys)
+
+[Service]
+Type=oneshot
+Environment=COREOS_METADATA_OPT_PROVIDER=--cmdline
+ExecStart=/usr/bin/coreos-metadata ${COREOS_METADATA_OPT_PROVIDER} --ssh-keys=%i
+
+[Install]
+DefaultInstance=core
```

Comment:
This embeds the assumption that there is a `core` user.  Perhaps we don't want to do that upstream?

Reply:
Good point.  Hmm, that definitely makes things harder.  We could add `make install-units USERNAME=core` or so?

Reply:
I think my vote for now is to move forward with the assumption FCOS-derived systems will also have a `core` user by default.

Reply:
FCOS-derived systems will have a `core` user, but this is going into the upstream package.  Maybe default the Makefile to `core`, with an override?

Reply:
I just checked, and `DefaultInstance=` can be overridden via a drop-in (which is properly honored at `enable` time). I'm not much of a fan of templating this here.

Reply:
Sure, but an arbitrary distro shouldn't have to (and, in practice, won't) ship a dropin if it wants to ship coreos-metadata.

Reply:
It seems to me like if we want to keep coreos-metadata as distro-agnostic as possible, the solution is templating in the user with the Makefile. I'm not too much of a fan of it as well, but unless someone has a different approach, I think we should just go with that for now.

Reply:
CT is not meant to be distro-agnostic though. It's _very_ specific to Container Linux (e.g. etcd, flannel, locksmith).

Reply:
ping on this discussion - I think it would be nice to get this in before we cut v2.0.0 (#93).

Reply:
After a bit of offline discussion with @sdemos, I'm not actually sure providing a `DefaultInstance` in the upstream unit makes sense at all. Non-{CL, RHCOS, FCOS} distros which may want to ship coreos-metadata are unlikely to be able to choose a single `DefaultInstance` that makes sense, and setting `DefaultInstance=core` via a dropin is simple enough for CL/RHCOS/FCOS.

Reply:
> Non-{CL, RHCOS, FCOS} distros which may want to ship coreos-metadata are unlikely to be able to choose a single DefaultInstance that makes sense

AFAIK the cloud image of pretty much every distro has a default user, whether it's `ec2-user`, `ubuntu`, `root`, etc.

Reply:
Ah, I suppose I was thinking mostly of non-cloud images, which doesn't really make sense in the context of coreos-metadata :slightly_smiling_face:

Reply:
@cgwalters @bgilbert @csssuf @lucab I think that given the constraints of the problem, templating is the best solution we have. Unless anyone is strongly opposed to it, I think @cgwalters should just implement it for this pr. If we come up with a better solution later, we can revisit this.

Reply:
> AFAIK the cloud image of pretty much every distro has a default user, whether it's ec2-user, ubuntu, root, etc.

Kind of...one really important detail wrt cloud-init is that that default user is only created if you don't specify another user.  Concretely for Fedora Atomic (and Fedora Cloud) if you boot a cloud-init config that looks like this (a reduced version of my default one):

```
users:
  - name: walters
    gecos: Colin Walters
    sudo: ALL=(ALL) NOPASSWD:ALL
    ssh-authorized-keys:
      - ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIOQkQHeKan3X+g1jILw4a3KtcfEIED0kByKGWookU7ev walters+2015-general@verbum.org
```

You *don't* get a `fedora` user.  Or to say it conversely, the user is only created dynamically on boot - and this is different from CL today where the `core` user exists by default.

Reply:
So any distribution that wants to use `coreos-metadata` would have to choose to either do what CL does and precreate that user, or alternatively have coreos-metadata do some sort of search dynamically for the "first" human user created by Ignition (possibly one that's in the `wheel/sudo` group?)

The thing that makes me uncomfortable about pre-creating e.g. a `core` user is - what if you join the node to LDAP/ActiveDirectory or whatever and there's already a "core" user?  You wouldn't want coreos-metadata granting ssh keys to that user...

In the end of course, an admin could delete the `core` user of course, but it does seem nicer to me to have the cloud-init semantics where it's a default user, but if you specify another one it doesn't exist.

Reply:
There's not a good way to express the "conditional default user" thing in Ignition right now; that'd be a good issue for coreos/ignition.

Note that we're talking about `coreos-metadata-sshkeys@.service`, not the entirety of coreos-metadata.  Distros which conditionally create the default user can't unconditionally enable that unit anyway, since they don't know what account to associate it with.  In that case, I think it makes sense for the distro to either

1. ignore the `DefaultInstance` and always specify the desired instance, or
2. set the `DefaultInstance` to the conditional default user, and only enable `-sshkeys@.service` when the default user is actually created.

Either way, setting `DefaultInstance` is either useful or harmless.

Reply:
Split that to https://github.com/coreos/ignition/issues/600